firewood/src/merkle.rs

```diff
@@ -71,11 +83,70 @@ impl<S: ShaleStore<Node>> Merkle<S> {
     }
 }
 
-impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
-    pub fn new(store: Box<S>) -> Self {
-        Self { store }
+impl<'de, S: ShaleStore<Node> + Send + Sync, T: BinarySerde> Merkle<S, T>
+where
+    EncodedNode<T>: serde::Deserialize<'de>,
+{
+    pub fn new(store: Box<S>, serde: T) -> Self {
+        Self { store, serde }
+    }
+
+    // TODO: use `get_encoded` / `decode` instead of `node.encode` / `node.decode` after extention node removal.
+    #[allow(dead_code)]
+    fn get_encoded(&self, node: &ObjRef) -> Result<Vec<u8>, MerkleError> {
+        let encoded = match node.inner() {
+            NodeType::Leaf(n) => EncodedNode {
+                node: EncodedNodeType::Leaf(n.clone()),
+                phantom: PhantomData,
+            },
+            NodeType::Branch(n) => {
+                let mut children: Box<[Option<Vec<u8>>; MAX_CHILDREN]> = Default::default();
+                for (pos, chd) in n.chd().iter().enumerate() {
+                    match chd {
+                        Some(chd_addr) => {
+                            let node = self.get_node(*chd_addr)?;
+
+                            let chd_encoded = self.get_encoded(&node)?;
+                            children[pos] = Some(chd_encoded);
+                        }
+                        None => {
+                            if let Some(v) = &n.children_encoded[pos] {
+                                children[pos] = Some(v.to_vec());
+                            }
+                        }
+                    }
+                }
+                EncodedNode {
+                    node: EncodedNodeType::Branch {
+                        children,
+                        value: n.value.clone(),
+                    },
+                    phantom: PhantomData,
+                }
+            }
+
+            NodeType::Extension(_) => todo!(),
+        };
+        Ok(self.serde.serialize(&encoded).unwrap())
+    }
+
+    #[allow(dead_code)]
+    fn decode(&self, buf: &'de [u8]) -> Result<NodeType, MerkleError> {
+        let encoded: EncodedNode<T> = self.serde.deserialize(buf).unwrap();
+        match encoded.node {
+            EncodedNodeType::Leaf(leaf) => Ok(NodeType::Leaf(LeafNode::new(leaf.path, leaf.data))),
+            EncodedNodeType::Branch { children, value } => {
+                let path = Vec::new().into();
+                let value = value.map(|v| v.0);
```

Comment:
🤔, you shouldn't need to do this, the branch is just going to wrap the `Vec` in `Data` again.